### ISSUE TITLE
hyperparameter scheduling

### ISSUE THREAD

**Issue by tetelias** [pytorch/xla#990]
Reading different issues I picked up that changing some of training loop parameters creates graph recalculation/time penalty. Will changing learning rate and momentum of the optimizer do that as well? Actually it's not even clear how to use a hyperparameter scheduler within training loop. Can you clarify?

**Comment by dlibenzi** [comment]
Unless you have Python control flow code which generates different graphs, you should not have any recompilation issue by changing the hyper-parameters.

**Comment by tetelias** [comment]
So I usually use this training loop in pytorch:
for x, (data, target) in loader:
    optimizer.zero_grad()
    output = model(data)
    loss = criterion(output, target)
    loss.backward()            
    optimizer.step()
    scheduler.step()

Do I just use the following in my TPU code?
for x, (data, target) in loader:
    optimizer.zero_grad()
    output = model(data)
    loss = loss_fn(output, target)
    loss.backward()
    xm.optimizer_step(optimizer)
    scheduler.step()

or 
for x, (data, target) in loader:
    optimizer.zero_grad()
    output = model(data)
    loss = loss_fn(output, target)
    loss.backward()
    scheduler.step()
    xm.optimizer_step(optimizer),

where scheduler changes lr or lr/momentum

**Comment by dlibenzi** [comment]
I don't know what your `scheduler.step()` does, but the latter is likely better.
The `xm.optimizer_step(optimizer)` sends a TPU execution in background, so that we can overlap that with the next step PyTorch code execution (which is essentially calling into ByTorch ops and we are building an IR graph behind).
But if we are hitting ops (after the `xm.optimizer_step(optimizer)` is sent) that requires TPU computations, we have to wait for the background one to complete.
So in general the latter should always be better or same, to the former.

**Comment by tetelias** [comment]
Thanks for clearing that up!

Also regarding batchnorm: its parameters are calculated separately on each core and there's no syncing, right? SyncBatchNorm is present in Pytorch, but it only works with DistributedDataParallel, which I haven't seen in XLA version.

**Comment by dlibenzi** [comment]
I am not familiar with `SyncBatchNorm`, but we do have fast reduce capabilities on TPU which could be used the synchronize parameters.
At the moment, I'd say no, `SyncBatchNorm` is likely not working (or poorly working if bounding on CPU).

**Comment by tetelias** [comment]
Still you just mentioned fast reduce capabilities on TPU. Where should I look for that? Any examples?

**Comment by dlibenzi** [comment]
So far we only expose Sum. Like:

https://github.com/pytorch/xla/blob/d0e5be7b968a6e363d496b1a9fbf1b5d495ef82f/torch_xla_py/xla_model.py#L369

If `a` is an XLA tensor, doing:
```Python
torch_xla._XLAC._xla_cross_replica_sum([a], scale, [])
# Every use of 'a' after this point will have Sum(a) * scale value
use(a, ...)
```

But in order for uses of reductions in places different from the optimizer step, we have also a few changes to do within our tensor core.

**Comment by tetelias** [comment]
Thanks, I'll take a look to see if can get anything from that!
